File: server/src/main/java/lof/research/lab/RecordSeederModule.java
package lof.research.lab;

import com.rpl.rama.*;
import com.rpl.rama.module.*;
import com.rpl.rama.ops.Ops;
import com.rpl.rama.test.*;
import lof.research.lab.data.Record;
import lof.research.lab.parsers.ZoteroXmlParser;

import java.util.List;

public class RecordSeederModule implements RamaModule {

  @Override
  public void define(Setup setup, Topologies topologies) {
    // Declare depot for Record ingestion
    setup.declareDepot("*recordsDepot", Depot.random());

    // Create microbatch topology for ETL processing
    MicrobatchTopology mb = topologies.microbatch("recordSeeder");

    // Declare PState: Map from canonicalId -> Record
    mb.pstate("$$recordsById", PState.mapSchema(String.class, Record.class));

    // Data flow: depot -> explode -> partition by ID -> store in PState
    mb.source("*recordsDepot").out("*microbatch")
      .explodeMicrobatch("*microbatch").out("*record")
      .each((Record r) -> r.canonicalId, "*record").out("*id")
      .hashPartition("*id")
      .localTransform("$$recordsById", Path.key("*id").termVal("*record"));
  }

  public static void main(String[] args) throws Exception {
    System.out.println("=== Record Seeder V0 ===");
    System.out.println("Parsing XML and seeding PState with first 10 records\n");

    // Parse XML file (first 10 records)
    String xmlPath = "../gsbbib__pretty.xml";
    List<Record> records = ZoteroXmlParser.parseXmlFile(xmlPath, 10);

    if (records.isEmpty()) {
      System.err.println("No records parsed. Exiting.");
      return;
    }

    // Launch in-process Rama cluster
    try (InProcessCluster cluster = InProcessCluster.create()) {
      RamaModule module = new RecordSeederModule();
      cluster.launchModule(module, new LaunchConfig(4, 4));
      String moduleName = module.getClass().getName();

      // Get depot and PState references
      Depot recordsDepot = cluster.clusterDepot(moduleName, "*recordsDepot");
      PState recordsById = cluster.clusterPState(moduleName, "$$recordsById");

      // Append all parsed records to depot
      System.out.println("Appending " + records.size() + " records to depot...");
      for (Record record : records) {
        recordsDepot.append(record);
      }

      // Wait for all records to be processed
      cluster.waitForMicrobatchProcessedCount(moduleName, "recordSeeder", records.size());
      System.out.println("All records processed!\n");

      // Query PState and display results
      System.out.println("=== Sample Records from PState ===\n");
      for (int i = 0; i < Math.min(3, records.size()); i++) {
        Record record = records.get(i);
        String id = record.canonicalId;
        Record retrieved = (Record) recordsById.selectOne(Path.key(id));

        if (retrieved != null) {
          System.out.println("Record " + (i + 1) + ":");
          System.out.println("  ID: " + retrieved.canonicalId);
          System.out.println("  Title: " + retrieved.title);
          System.out.println("  Authors: " + retrieved.authors);
          System.out.println("  Year: " + retrieved.year);
          System.out.println("  DOI: " + (retrieved.doi != null ? retrieved.doi : "N/A"));
          System.out.println("  URL: " + (retrieved.url != null ? retrieved.url : "N/A"));
          System.out.println();
        } else {
          System.err.println("ERROR: Could not retrieve record with ID: " + id);
        }
      }

      System.out.println("=== Summary ===");
      System.out.println("Total records in PState: " + records.size());
      System.out.println("V0 seeder complete!");
    }
  }
}

File: server/src/main/clj/lof/lab/record_seeder_module.clj
(ns lof.lab.record-seeder-module
  (:use [com.rpl.rama]
        [com.rpl.rama.path])
  (:require [com.rpl.rama.ops :as ops]
            [lof.lab.data.record :as rec])
  (:import [lof.research.lab.parsers ZoteroXmlParser]
           [lof.lab.data.record Record]))

;; This module demonstrates XML ingestion -> PState seeding for academic papers.
;; It takes parsed Record objects from a Zotero XML export and stores them in a
;; PState indexed by canonical-id.

(defmodule RecordSeederModule
  [setup topologies]
  ;; Depot for Record ingestion, partitioned by canonical-id
  (declare-depot setup *records-depot (hash-by :canonical-id))

  ;; Microbatch topology for ETL processing
  (let [mb (microbatch-topology topologies "record-seeder")]
    ;; PState: Map from canonical-id (String) -> Record
    (declare-pstate mb $$records-by-id {String Record})

    ;; Dataflow: depot -> explode batch -> partition by ID -> store in PState
    (<<sources mb
      ;; Subscribe to depot, bind microbatch
      (source> *records-depot :> %microbatch)

      ;; Explode batch and destructure Record fields
      (%microbatch :> {:keys [*canonical-id] :as *record})

      ;; Partition by canonical-id
      (|hash *canonical-id)

      ;; Write to PState
      (local-transform> [(keypath *canonical-id) (termval *record)]
                        $$records-by-id))))

(defn -main
  "Parse XML, seed depot, and query PState"
  [& args]
  (println "=== Record Seeder V0 (Clojure) ===")
  (println "Parsing XML and seeding PState with first 10 records\n")

  ;; Parse XML file using Java interop (first 10 records)
  (let [xml-path "../gsbbib__pretty.xml"
        java-records (ZoteroXmlParser/parseXmlFile xml-path 10)
        ;; Convert Java Records to Clojure Records
        records (mapv (fn [jr]
                        (rec/->Record
                          (.canonicalId jr)
                          (vec (.authors jr))
                          (.title jr)
                          (.year jr)
                          (.doi jr)
                          (.url jr)))
                      java-records)]

    (if (empty? records)
      (println "No records parsed. Exiting.")

      ;; Launch in-process Rama cluster
      (with-open [ipc (com.rpl.rama.test.InProcessCluster/create)]
        (com.rpl.rama.test.InProcessCluster/launchModule
          ipc RecordSeederModule
          (com.rpl.rama.test.LaunchConfig. 4 4))

        (let [module-name (get-module-name RecordSeederModule)
              records-depot (foreign-depot ipc module-name "*records-depot")
              records-by-id (foreign-pstate ipc module-name "$$records-by-id")]

          ;; Append all parsed records to depot
          (println (str "Appending " (count records) " records to depot..."))
          (doseq [record records]
            (foreign-append! records-depot record))

          ;; Wait for all records to be processed
          (com.rpl.rama.test.InProcessCluster/waitForMicrobatchProcessedCount
            ipc module-name "record-seeder" (count records))
          (println "All records processed!\n")

          ;; Query PState and display results
          (println "=== Sample Records from PState ===\n")
          (doseq [i (range (min 3 (count records)))]
            (let [record (nth records i)
                  id (:canonical-id record)
                  retrieved (foreign-select-one (keypath id) records-by-id)]

              (if retrieved
                (do
                  (println (str "Record " (inc i) ":"))
                  (println (str "  ID: " (:canonical-id retrieved)))
                  (println (str "  Title: " (:title retrieved)))
                  (println (str "  Authors: " (:authors retrieved)))
                  (println (str "  Year: " (:year retrieved)))
                  (println (str "  DOI: " (or (:doi retrieved) "N/A")))
                  (println (str "  URL: " (or (:url retrieved) "N/A")))
                  (println))
                (println (str "ERROR: Could not retrieve record with ID: " id)))))

          (println "=== Summary ===")
          (println (str "Total records in PState: " (count records)))
          (println "V0 seeder complete!"))))))

(comment
  ;; REPL testing
  (-main)

  ;; Or manual step-by-step:
  (require '[com.rpl.rama.test :as rtest])

  (def ipc (rtest/create-ipc))
  (rtest/launch-module! ipc RecordSeederModule {:tasks 4 :threads 2})

  (def module-name (get-module-name RecordSeederModule))
  (def depot (foreign-depot ipc module-name "*records-depot"))
  (def pstate (foreign-pstate ipc module-name "$$records-by-id"))

  ;; Append a test record
  (foreign-append! depot
                   (rec/make-record {:authors ["Test, Author"]
                                     :title "Test Title"
                                     :year 2024
                                     :doi nil
                                     :url nil}))

  ;; Query
  (foreign-select-one (keypath "test-2024-test-title") pstate)

  (.close ipc)
  )
